### ISSUE TITLE
Add [toc] support

### ISSUE THREAD

**Issue by Evrey** [azerupi/mdBook#153]
Add support for the `[toc]` tag. How does it work? It takes every level 2-6 header and adds a table of contents for them. I.e., this markdown here...

```
# A
[toc]

## A.b
test

### A.b.d
moah test

## A.c
test
```

... would be rendered as something similar to this:

# A

- [A.b](#test-a-b)
    - [A.b.d](#test-a-b-d)
- [A.c](#test-a-c)

## A.b
test

### A.b.d
moah test

## A.c
test

Obviously, those links should jump to anchors near the sub-chapter headers. This is an IMHO really neat feature to have. Right now I'm emulating that stuff with lots of annoying `<a name="some-anchor"></a>` before each header.

**Comment by azerupi** [comment]
A table of contents for pages has been requested previously I think. 

I personally dislike the idea of a TOC per chapter (especially inline with the content) as it would only be useful for chapters that are quite long. And if that's the case I would recommend using sub-chapters. But my opinion is not everyone's opinion, so if this is a wanted feature it should be available in some form. 

Is this `[toc]` tag already used somewhere?

**Comment by Evrey** [comment]
Yeah, it is a somewhat-known Markdown extension, like tables are GitHub extensions. E.g. the Markdown editor Haroopad supports `[toc]`. And Python Markdown as I've been told.

The thing is: I'm using `mdbook` for quite heavy design documents which already have a lot of chapters and sub-chapters in them. Just to give you an example:

One of my `mdbook` projects is the design document of a story-driven game. There are already a bunch of larger chapters like gameplay, story, general character sheets, and what not. Right inside those chapters there are files per main character, which only contain information about the corresponding topics. I.e. the character files in story don't need stuff about what the characters look like or about what skill sets they have. Still, those character fies become quite long and densely packed with information, and splitting them off into yet even more sub-chapters... well... at some point I don't see the potential value besides having yet even more tabs opened simultaneously to read all the information I need.

`[toc]` only creates a table of contents per file and even if you were to include `[toc]`, nothing will hold anyone back from not using it.

There is, however, one single issue I see with including `[toc]` (`[TOC]` is also valid): The print page.

I think, that `[toc]` would pretty much fuck up the print page, so this tag could/should be disabled while rendering it?

**Comment by azerupi** [comment]
[This issue](https://github.com/azerupi/mdBook/issues/58) asks for a similar feature. 

I think there could indeed be some value to add such a feature, I am however not sure how to best implement it.

I like the MDN example better, where the toc is on the right side instead of inline with the content. It would allow the toc to stay visible when you scroll. But it would take some screen estate and make everything a little denser. I would also see some value for the user to enable it automatically, possibly only for long chapters, instead of having the author add it manually for each page. What do you think?

**Comment by Evrey** [comment]
Sounds interesting indeed, but having it on the right will use way too much space, I guess. How about putting the local TOC in the same column as the summary? Split it vertically in half. The local TOC shouldn't require that much of vertical space. Just a suggestion.

**Comment by azerupi** [comment]
**Note to myself:** In [#268 (comment)](https://github.com/azerupi/mdBook/issues/268#issuecomment-301700191) I mentioned that we should probably find a common syntax for mdBook extensions in the source that is not going to collide with CommonMark.

**Comment by dwijnand** [comment]
> [This issue](https://github.com/rust-lang-nursery/mdBook/issues/58) asks for a similar feature.

maybe best close this one marked as duplicate, and leave a comment on that issue summarising the details here (or just mentioning it)?

**Comment by Michael-F-Bryan** [comment]
I'm in two minds about this. I've got the bad habit of explaining things in way too much detail (to quote Pascal, "*If I Had More Time, I Would Have Written a Shorter Letter*") and for really long articles I would really like some sort of `[toc]` tag to automatically insert a table of contents. 

But then you also fall into the trap of making mdbook-specific markdown extensions, which is both bad for complexity and bad for flexibility (e.g. our `[toc]` probably wouldn't work when viewed on GitHub). Not having the ability to generate tables of contents also forces people to keep things short and manageable, and use sub-chapters to break things up. I don't think we should be forcing style choices on users though...

**Comment by AndydeCleyre** [comment]
I'm an outsider here on multiple counts: 
- I rarely develop in rust
- wherever possible, I use reStructuredText over Markdown
- I tend to have small projects with all documentation in a single README

That said, mdBook output is beautiful, and this is the only issue stopping me from using it instead of Sphinx/Read the Docs, so I'm going to add my use case here.

Each project has a template `README.rst.t`, which my build scripts render into a `README.rst` for general/GitHub rendering, and an `index.rst` for Sphinx/rtd rendering. The only difference between those two outputs is that the in-body TOC is excluded from `index.rst`, to let the Sphinx theme handle the navigation in its own way (and the depth level is off-by-one, between Sphinx and plain/GitHub). This is the TOC in the template:

```rst
<!--(if default("include_toc", True))-->
.. contents::
   :depth: 1
<!--(end)-->
```

I began adding an mdBook build of the readme to a project, but am hung up on this issue, as the large sidebar only has one item, where I'd like it to at least have the top-level sections of the current (only) page, and maybe also some arbitrary links. I thought maybe playing with the `[output.html.fold]` options could get me to "unwrap" the current page, but no.

Even if making use of the sidebar for this is impossible, I'd hope to get an in-body TOC, like GitHub will render when I include it explicitly (and then I'd see if I could disable or auto-hide the sidebar altogether).

As I saw CommonMark mentioned in an issue that mentioned reStructuredText, I'm using this to generate the Markdown:

```console
$ pandoc /path/to/README.rst --to commonmark -o src/README.md
```

The final render is very good, aside from the lack of in-page navigation. The Markdown generated by that `pandoc` command does include:

```html
<div class="contents" data-depth="1">

</div>
```

But no in-body TOC is present in the final render. I'll have to take a closer look, maybe there's another of the many Markdown "standards" that mdBook more closely follows. I don't know HTML or Markdown well enough to know if that snippet is proper anyway.

**Comment by mgeier** [comment]
@AndydeCleyre 

> mdBook output is beautiful, and this is the only issue stopping me from using it instead of Sphinx/Read the Docs

I've created a Sphinx theme inspired by `mdBook` an I've released it just a few minutes ago: https://insipid-sphinx-theme.readthedocs.io/.

It's not an exact clone of `mdBook`, but it looks similar, so maybe it suits your needs.

Please let me know what you think about it!

**Comment by AndydeCleyre** [comment]
Thanks @mgeier! The colors are not friendly for my eyes. But in following up I noticed #1263, which led me to [JorelAli/mdBook-pagetoc](https://github.com/JorelAli/mdBook-pagetoc), which may be a good solution for me; I'll try it. I may then see about disabling the usual sidebar completely in my builds.

**Comment by kalindkaria** [comment]
@AndydeCleyre I would suggest to use the preprocessor `mdbook-toc` implemented by @badboy.
It really serves the purpose.

**Comment by Michael-F-Bryan** [comment]
Should this be closed because it has been solved by the ecosystem?

**Comment by kalindkaria** [comment]
Yes, I think.

**Comment by sanmai-NL** [comment]
@Evrey Are you ready to close this PR?
